src/stream/src/cache/range_lru.rs

```diff
@@ -0,0 +1,542 @@
+        {
+            let encoded_range = (
+                translate_lower_bound(range.0.clone()),
+                translate_upper_bound(range.1.clone()),
+            );
+
+            let mut iter = self.secondary_index.range(encoded_range).peekable();
+
+            while let Some((k, v)) = iter.next() {
+                remove_keys.push(k.clone());
+                // Delete all of these, replace it with our range. Keep track of the last deleted.
+                // Exceptions:
+                if iter.peek().is_some() {
+                    // Not the last, so the epoch must be `Some`
+                    self.last_accessed_at.remove(v.as_ref().unwrap());
+                } else {
+                    if let Some(e) = v {
+                        // Our secondary index contains this epoch, so it must be `Some`
+                        {
+                            let split_range = self.last_accessed_at.get_mut(e).unwrap();
+                            if split_range.1 == range.1 {
+                                delete_upper_range = true;
+                            } else {
+                                split_range.0 = match range.1.clone() {
+                                    Excluded(x) => Included(x),
+                                    Included(x) => Excluded(x),
+                                    _ => unreachable!(),
+                                };
+                            }
+                        }
+                        if delete_upper_range {
+                            self.last_accessed_at.remove(e).unwrap();
+                        }
+                    }
+                    last_deleted_epoch = *v;
+                }
+            }
+            // Deal with the precursor to the range (by shortening it).
+            if !matches!(range.0, Unbounded) {
+                let upper = match translate_lower_bound(range.0.clone()) {
+                    Included(x) => Excluded(x),
+                    _ => unreachable!(),
+                };
+                let lower_range = (Unbounded, upper);
+                let mut rev_iter = self.secondary_index.range(lower_range).rev();
+
+                let e = rev_iter
+                    .next()
+                    .map_or(self.lower_unbounded_epoch, |(_, &x)| x);
+                if let Some(e) = e {
+                    let split_range = self.last_accessed_at.get_mut(&e).unwrap();
+                    split_range.1 = match range.0.clone() {
+                        Excluded(x) => Included(x),
+                        Included(x) => Excluded(x),
+                        _ => unreachable!(),
+                    };
+                }
+            }
+        }
+        for key in remove_keys {
+            self.secondary_index.remove(&key).unwrap();
+        }
+        self.last_accessed_at.insert(epoch, range.clone());
+
+        // Insert lower bound of our range into secondary index
+        if matches!(range.0, Unbounded) {
```

Comment:
If we have `lower_unbounded_epoch` for this epoch, shall we also remove the smallest key in the secondary index with the same epoch?

Reply:
I believe this would have been removed in `remove_keys` if as that bound would overlap with the current range.